Allow Column `select` to support logical buffers

ezmiller opened the pull request:
# Goal

Allow tech.ml.dataset's column `select` method to support logical buffers. This would make this kind of convenient selection available, which _I think_ may not currently be possible:

```clj
  (def x (new-column [1 nil 3 4 5 6 7]))

  x

  (.select x (fun/not (fun/nan? x)))
  ;; => #tech.v3.dataset.column<int64>[6]
  ;;    null
  ;;    [1, 3, 4, 5, 6, 7]

  (.select x (fun/> x 5))
  ;; => #tech.v3.dataset.column<int64>[2]
  ;;    null
  ;;    [6, 7]
```
This is convenient, and it will also be familiar to users of R and Python who can do this:
```r
```

cnuernber commented on src/tech/v3/dataset/impl/column.clj:
```diff
@@ -329,7 +329,9 @@
   (select [col idx-rdr]
-    (let [idx-rdr (->efficient-reader idx-rdr)]
+    (let [idx-rdr (if (= (dtype/elemwise-datatype idx-rdr) :boolean)
+                    (->efficient-reader (keep-indexed #(when (true? %2) %1) idx-rdr))
```
I think there may better way to do this - 
[bool-reader->indexes](https://github.com/cnuernber/dtype-next/blob/master/src/tech/v3/datatype/unary_pred.clj#L267).  You will have to ensure it is a reader before calling that method - [ensure-reader](https://cnuernber.github.io/dtype-next/tech.v3.datatype.argops.html#var-ensure-reader).

It is conceptually the same as [argfilter](https://cnuernber.github.io/dtype-next/tech.v3.datatype.argops.html#var-argfilter) - `(argops/argfilter identity idx-rdr).

ezmiller replied:
Oh nice. I was wondering if there was a more efficient way to do this.

ezmiller commented:
@cnuernber one other funny thing here is this if a user were to try to use a regular clojure vector:
```clj
(def c (col/new-column [1 2 3 4 5]))

(col/select c (mapv (partial < 3) c))
;; => #tech.v3.dataset.column<int64>[5]
;;    null
;;    [1, 1, 1, 2, 2]
```
This wouldn't happen if they use `tech.v3.datatype.functional`, but you could imagine it happening if the user tried to use plain clojure functions.

It doesn't seem like a big problem for this particular PR, but I mention it in part because I feel like I run into this kind of issue frequently. I've wondered if there is a way to smooth it out. Maybe there isn't? Because conceptually the Clojure vector is not like a dtype-next buffer, i.e. it is not typed homogeneously? 

Practically speaking you'd have to scan the vector somehow to see if it can be cast to a logical buffer? I wonder if there's a way to do this efficiently and know when to do it?

ezmiller commented:
@cnuernber I switched to use `bool-reader->indexes` and adjusted dataset `select` to parse the logical buffer at that higher level. 

Re: scanning, I have only a sketchy idea of what it means to "scan". I once followed out the process that happens from `new-column` and the functions in the `tech.v3.dataset.impl.column-data-process namespace`, which end up auto-detecting the type contained in a generic vector.

That's what I was imagining: that we need to know the type is boolean so that we know to build an array of indexes from the positions of true values, i.e. the distinction that the code in this PR is doing so far. 

But I think you are saying something a bit different? When you say scan in this context where the regular clojure vector in question is an input to a column or dataset `select` fn call, what does scanning mean? Where is it happening? Do you mean what's happening inside e.g. `->efficient-reader` that at a low level it is scanning the vector, so we could just write something that does this check for truthy (or actually just true/boolean?) values on the fly?
